### PR TITLE
New Resource: aws_s3control_bucket_lifecycle_configuration

### LINKED ISSUE #15415: New Resource: aws_s3control_bucket_lifecycle_configuration
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As part of the [initial S3 on Outposts support for the Terraform AWS Provider](https://github.com/terraform-providers/terraform-provider-aws/issues/15412), a new resource for managing Bucket Lifecycle Configurations in the S3 Control API is necessary. This new resource does not deprecate the existing `aws_s3_bucket` resource functionality and does not represent any operator need for migration between the two. Resource documentation call out boxes will be in place to help delineate the purpose of the two similar, but different resources.

### New Resource

* aws_s3control_bucket_lifecycle_configuration

### Potential Terraform Configuration

```hcl
resource "aws_s3control_bucket_lifecycle_configuration" "example" {
  bucket = aws_s3control_bucket.example.bucket

  rule {
    abort_incomplete_multipart_upload {
      days_after_initiation = 123
    }

    expiration {
      date                         = ""
      days                         = 123
      expired_object_delete_marker = true
    }

    filter {
      and {
        prefix = ""
        tag {
          key   = ""
          value = ""
        }
      }
      prefix = ""
      tag {
        key   = ""
        value = ""
      }
    }
    id = ""
    non_current_version_expiration {
      non_current_days = 123
    }
    non_current_version_transition {
      non_current_days = 123
      storage_class    = ""
    }
    transition {
      date          = ""
      days          = 123
      storage_class = ""
    }
  }
}
```

### References

* [Feature Request: S3 on Outposts Support](https://github.com/terraform-providers/terraform-provider-aws/issues/15412)
* [New Resource: aws_s3control_bucket](https://github.com/terraform-providers/terraform-provider-aws/issues/15413)
* [Update module aws/aws-sdk-go to v1.35.0](https://github.com/terraform-providers/terraform-provider-aws/pull/15226)
* [S3 Control API Reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Operations_AWS_S3_Control.html)
* [AWS Go SDK v1.35.0](https://github.com/aws/aws-sdk-go/releases/tag/v1.35.0)
* [AWS Go SDK service/s3control](https://docs.aws.amazon.com/sdk-for-go/api/service/s3control/)

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Reference: https://github.com/aws/aws-sdk-go/issues/3591
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/15412
Requires #15468
Requires #15510
Closes #15415

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource:** `aws_s3control_bucket_lifecycle_configuration`
```

Output from acceptance testing (note everything except AbortIncompleteMultipartUpload was tested before losing Outpost access):

```
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_basic (59.06s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_disappears (59.96s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Expiration_Date (70.35s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Expiration_Days (57.63s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Filter_Prefix (64.31s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Filter_Tags (62.22s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Id (73.91s)
--- PASS: TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Status (65.70s)
```

### CODE REVIEW

File: aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+						"expiration.0.days": "30",
+					}),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSS3ControlBucketLifecycleConfiguration_Rule_Expiration_ExpiredObjectDeleteMarker(t *testing.T) {
+	TestAccSkip(t, "S3 on Outposts does not error or save it in the API when receiving this parameter")
```
Comment (anGie44) [inline aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:194]: given this behavior - should this be documented in the resource docs for users?

File: aws/resource_aws_s3control_bucket_lifecycle_configuration.go [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+
+											_, err := time.Parse("2006-01-02", value)
+
+											if err != nil {
+												errors = append(errors, fmt.Errorf("%q should be in YYYY-MM-DD date format", value))
+											}
+
+											return
+										},
+									},
```
Comment (anGie44) [inline aws/resource_aws_s3control_bucket_lifecycle_configuration.go:74]: could also live in `validators.go` ? or possibly getting too large to keep populating ?

**Review by anGie44 — APPROVED** [review]
LGTM -- just a couple small q's and noting updating an existing `tags` key:value pair may be affected by the open issue in the AWS SDK(https://github.com/aws/aws-sdk-go/issues/3591)

File: aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+  bucket     = %[1]q
+  outpost_id = data.aws_outposts_outpost.test.id
+}
+
+resource "aws_s3control_bucket_lifecycle_configuration" "test" {
+  bucket = aws_s3control_bucket.test.arn
+
+  rule {
+    expiration {
+      days                         = %[2]t ? null : 365
```
Comment (anGie44) [inline aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:588]: i think this line is causing the terrafmt error: `./aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:571 failed to process with: failed to parse hcl: ./aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:17,38-39: Invalid expression; Expected the start of an expression, but found an invalid expression token."`

Comment (anGie44) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:588]: maybe `terrafmt` still needs to support ternaries w/formatters?

File: aws/resource_aws_s3control_bucket_lifecycle_configuration.go [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+
+	return tfMaps
+}
+
+func flattenS3controlLifecycleRule(apiObject *s3control.LifecycleRule) map[string]interface{} {
+	if apiObject == nil {
+		return nil
+	}
+
+	log.Printf("[BFLAD] flattenS3controlLifecycleRule() apiObject: %+v", apiObject)
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration.go:476]: haha nice 😅 

```suggestion
```

File: aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go [hunk 5]
```diff
@@ -0,0 +1,750 @@
...
+			return nil
+		}
+
+		if tfawserr.ErrCodeEquals(err, "NoSuchLifecycleConfiguration") {
+			return nil
+		}
+
+		if tfawserr.ErrCodeEquals(err, "NoSuchOutpost") {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:439]:
```suggestion
		if tfawserr.ErrCodeEquals(err, "NoSuchBucket") {
			continue
		}

		if tfawserr.ErrCodeEquals(err, "NoSuchLifecycleConfiguration") {
			continue
		}

		if tfawserr.ErrCodeEquals(err, "NoSuchOutpost") {
			continue
		}
```

File: aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go (hunk 3 above)
Comment (bflad) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:588]: We recently bumped `terrafmt` dependency on master and rebasing this pull request to fix the merge conflict fixed this. 👍

File: aws/resource_aws_s3control_bucket_lifecycle_configuration.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration.go:74]: Partially related to https://github.com/terraform-providers/terraform-provider-aws/issues/8424 -- although in this case its something that would ultimately be a good candidate for there (or a similar shared location) if it had more usage than just here and the `aws_s3_bucket` resource. 👍

File: aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go (hunk 1 above)
Comment (bflad) [reply aws/resource_aws_s3control_bucket_lifecycle_configuration_test.go:194]: Hopefully my out of band answer sufficed here, but we can certainly add this post merge if necessary. 👍

**Pull request merged**